File: 2020/2/both.py
#!/usr/bin/python

import re
import string
import sys

infile = 'input.txt'
if (len(sys.argv) > 1):
	infile = sys.argv[1]

fh = open(infile, 'r');

count1 = 0
count2 = 0

r = re.compile(r"^(\d+)-(\d+) ([a-z]): ([a-z]+)")

for line in fh:
	m = r.match(line)
	if m != None:
		a = int(m.group(1))
		b = int(m.group(2))
		letter = m.group(3)
		password = m.group(4)
		num = password.count(letter)
		if num >= a and num <= b:
			count1 += 1 
		c = password[a - 1]
		d = password[b - 1]
		if (c == letter) != (d == letter):
			count2 += 1

print("Part 1 answer: %d" % count1)
print("Part 2 answer: %d" % count2)

File: 2020/2/both.pl
#!/usr/bin/perl

use strict;
use integer;

my $fh;

die "* no input\n" if (!open($fh, '<', 'input.txt'));

my $count1 = 0;
my $count2 = 0;

while (my $line = <$fh>)
{
	if ($line =~ /^(\d+)-(\d+) ([a-z]): ([a-z]+)/) {
		my ($a, $b, $letter, $password) = ($1, $2, $3, $4);
		my $num = () = $password =~ /$letter/g;
		$count1++ if ($num >= $a && $num <= $b);
		my $c = substr($password, $a - 1, 1);
		my $d = substr($password, $b - 1, 1);
		$count2++ if (($c eq $letter) != ($d eq $letter));
	}
}

printf("Answer to part 1: %d\n", $count1);
printf("Answer to part 2: %d\n", $count2);
